inc/class-job.php

```diff
@@ -186,39 +194,184 @@ public static function get_by_site( $site, $include_completed = false, $include_
-			$table = static::get_table();
+		if ( empty( $results ) ) {
+			return [];
+		}
 
-			$sql = "SELECT * FROM `{$table}` WHERE site = %d";
-			$sql .= ' AND status IN(' . implode( ',', array_fill( 0, count( $statuses ), '%s' ) ) . ')';
-			$query = $wpdb->prepare( $sql, array_merge( [ $site ], $statuses ) );
-			$results = $wpdb->get_results( $query );
+		return static::to_instances( $results );
+	}
 
-			if ( ! $include_completed && ! $include_failed ) {
-				wp_cache_set( 'jobs', $results, 'cavalcade-jobs' );
-			}
+	/**
+	 * Query jobs database.
+	 *
+	 * Returns an array of Job instances for the current site based
+	 * on the paramaters.
+	 *
+	 * @todo: allow searching within time window for duplicate events.
+	 *
+	 * @param array|\stdClass $args {
+	 *     @param string          $hook      Jobs hook to return. Optional.
+	 *     @param int|string|null $timestamp Timestamp to search for. Optional.
+	 *                                       String shortcuts `future`: > NOW(); `past`: <= NOW()
+	 *     @param array           $args      Cron job arguments.
+	 *     @param int|object      $site      Site to query. Default current site.
+	 *     @param array           $statuses  Job statuses to query. Default to waiting and running.
+	 *     @param int             $limit     Max number of jobs to return. Default 1.
+	 *     @param string          $order     ASC or DESC. Default ASC.
+	 * }
+	 * @return Job[]|WP_Error Jobs on success, error otherwise.
+	 */
+	public static function get_jobs_by_query( $args = [] ) {
+		global $wpdb;
+		$args = (array) $args;
+		$results = [];
+
+		$defaults = [
+			'timestamp' => null,
+			'hook' => null,
+			'args' => [],
+			'site' => get_current_blog_id(),
+			'statuses' => [ 'waiting', 'running' ],
+			'limit' => 1,
+			'order' => 'ASC',
+			'__raw' => false,
+		];
+
+		$args = wp_parse_args( $args, $defaults );
+
+		// Allow passing a site object in
+		if ( is_object( $args['site'] ) && isset( $args['site']->blog_id ) ) {
+			$args['site'] = $args['site']->blog_id;
 		}
 
-		if ( empty( $results ) ) {
-			return [];
+		if ( ! is_numeric( $args['site'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_site_id' );
+		}
+
+		if ( ! empty( $args['hook'] ) && ! is_string( $args['hook'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_hook_name' );
+		}
+
+		if ( ! is_array( $args['args'] ) && ! is_null( $args['args'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_event_arguments' );
+		}
+
+		if ( ! is_numeric( $args['limit' ] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_limit' );
+		}
+
+		if ( $args['timestamp'] === 'future' ) {
+			$timestamp = time();
+			$timestamp_compare = '>';
+		} elseif ( $args['timestamp'] === 'past' ) {
+			$timestamp = time();
+			$timestamp_compare = '<=';
+		}
+
+		$args['limit' ] = absint( $args['limit' ] );
+
+		if ( $args['limit'] > 100 ) {
+			trigger_error( 'Exceeding recommended job search limit of 100' );
```

Comment:
Do we need this?

Reply:
Not sure, can remove in favour of the filter though. I'll add a note about maximum limits in the filter comment.

Reply:
100 rows from the DB should be fine I think (WP's limit is because 100 posts is more like 1000 objects once you include things like meta); I'd imagine 1000 items would be a sensible query. In any case, I'd say too opinionated for the core code here.